src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1782,10 +1787,12 @@ MergeTreeData::DataPartsVector MergeTreeData::grabOldParts(bool force)
                 || isInMemoryPart(part)     /// Remove in-memory parts immediately to not store excessive data in RAM
                 || (part->version.creation_csn == Tx::RolledBackCSN && getSettings()->remove_rolled_back_parts_immediately))
             {
+                part->removal_state.store(DataPartRemovalState::REMOVED, std::memory_order_relaxed);
                 parts_to_delete.emplace_back(it);
             }
             else
             {
+                part->removal_state.store(DataPartRemovalState::NOT_REACHED_REMOVAL_TIME, std::memory_order_relaxed);
```

Comment:
It also can mean that `has_skipped_mutation_parent(part)` was true (yep, the condition above is quite complex)